File: be/models/productVariant.model.js
const db = require("../configs/db");

const productVariantSchema = new db.mongoose.Schema(
  {
    product_id: {
      type: db.mongoose.Schema.Types.ObjectId,
      ref: "productModel",
      required: true,
    },
    variant_name: { type: String, required: true },
    variant_slug: { type: String, required: true },
    image: { type: String, default: null },
    price: { type: Number, required: true },
    sale_price: { type: Number, default: null },
    stock_quantity: { type: Number, required: true, default: 0 },
    sold_quantity: { type: Number, required: true, default: 0 },
    status: {
      type: String,
      required: true,
      enum: ["active", "inactive", "out_of_stock"],
      default: "active",
    },
    deleted_at: { type: Date, default: null },
  },
  { collection: "product_variants", timestamps: true },
);

let productVariantModel = db.mongoose.model(
  "productVariantModel",
  productVariantSchema,
);
module.exports = { productVariantModel };

File: tuhubread/lib/models/product_variant.model.dart
import 'package:json_annotation/json_annotation.dart';

part 'product_variant.model.g.dart';

@JsonSerializable(fieldRename: FieldRename.snake)
class ProductVariantModel {
  @JsonKey(name: '_id')
  final String id;
  final String productId;
  final String variantName;
  final String variantSlug;
  final String? image;
  final double price;
  final double? salePrice;
  final int stockQuantity;
  final int soldQuantity;
  final String status;

  ProductVariantModel({
    required this.id,
    required this.productId,
    required this.variantName,
    required this.variantSlug,
    this.image,
    required this.price,
    this.salePrice,
    required this.stockQuantity,
    required this.soldQuantity,
    required this.status,
  });

  factory ProductVariantModel.fromJson(Map<String, dynamic> json) =>
      _$ProductVariantModelFromJson(json);

  Map<String, dynamic> toJson() => _$ProductVariantModelToJson(this);
}
